actors/builtin/miner/miner_state.go

```diff
@@ -309,46 +301,185 @@ func (st *State) ForEachSector(store adt.Store, f func(*SectorOnChainInfo)) erro
 
 // Adds some sector numbers to the new sectors bitfield.
 func (st *State) AddNewSectors(newSectors ...abi.SectorNumber) error {
-	panic("Implement me") // TODO
+	for _, sector := range newSectors {
+		st.NewSectors.Set(uint64(sector))
+	}
+
+	return nil
 }
 
 // Clears the new sectors bitfield.
 func (st *State) ClearNewSectors() error {
-	panic("implement me") // TODO
+	st.NewSectors = abi.NewBitField()
+	return nil
 }
 
 // Adds some sector numbers to the set expiring at an epoch.
 func (st *State) AddSectorExpirations(store adt.Store, expiry abi.ChainEpoch, newSectors ...abi.SectorNumber) error {
-	panic("implement me") // TODO
+	arr, err := adt.AsArray(store, st.SectorExpirations)
+	if err != nil {
+		return err
+	}
+
+	bf := abi.NewBitField()
+	_, err = arr.Get(uint64(expiry), &bf)
+	if err != nil {
+		return err
+	}
+
+	for _, sector := range newSectors {
+		bf.Set(uint64(sector))
+	}
+
+	if err := arr.Set(uint64(expiry), &bf); err != nil {
+		return err
+	}
+
+	st.SectorExpirations, err = arr.Root()
+	return err
+}
+
+func checkBitFieldIntersect(expectAllIntersect bool, a, b abi.BitField) (bool, uint64, error) {
+	// TODO: bitField.Intersects(other)
+	am, err := a.AllMap(MaxFaultsCount)
+	if err != nil {
+		return false, 0, err
+	}
+
+	bm, err := b.All(MaxFaultsCount)
+	if err != nil {
+		return false, 0, err
+	}
+
+	for _, i := range bm {
+		if ok := am[i]; ok != expectAllIntersect {
+			return false, i, nil
+		}
+	}
+
+	return true, 0, nil
 }
 
 // Adds sectors numbers to faults and fault epochs.
 // Precondition: sectors are not already in faults.
-func (st *State) AddFaults(sectorNos abi.BitField, faultEpoch abi.ChainEpoch) error {
-	panic("implement me") // TODO
+func (st *State) AddFaults(store adt.Store, sectorNos abi.BitField, faultEpoch abi.ChainEpoch) (err error) {
+	// TODO: check max count after adding
+
+	{
+		ok, bad, err := checkBitFieldIntersect(false, st.Faults, sectorNos)
```

Comment:
Idea: rather than expand maps for this check, you could do the merge and check the expected size is the sum of sizes.